Add native user authorization, login, and clan management

copperwater opened the pull request:
as well as a couple other minor updates and additions, but this is the main thing.

copperwater committed:
Fix freenode link

copperwater committed:
Add clan message, clan admin, and clan invites to models

copperwater committed:
Add small script to automate work of un-migrating and re-migrating

Which is a common thing to want to do when the project is new and you
don't have database compatibility with previous versions to worry about.

Running this will wipe all migrations and create a new 0001_initial
migration, based off the current state of models.py.

copperwater committed:
Add native user authorization, login, and clan management

Moving away from the old scoreboard's approach, where "clan management"
took you to a completely different site in which you did clan things.
This approach worked fine if the clan database was the only true
database, and the entire rest of the site was statically generated based
on its contents plus xlogfile contents. But that's not the type of site
we're trying to build now. Clans and players are an integral part of
the database, and keeping it an arm's length away didn't make sense.

The result: the clan management and login system, rewritten from the
ground up in Django. I've tried to match functionality if not appearance
with the old system. Namely: anyone can make a clan, becoming the admin
of that clan; they can then invite people to the clan and manage
membership by bestowing admin rights, making and rescinding invites, and
kicking members out if necessary. The frontend styling isn't top-notch,
but the data is there, it's functional, and it doesn't look like a
complete wreck, so this is no time to be perfectionist about it.

This also tweaks a couple other things; I added a TODO here and there
when I noticed something to be done, and as an experiment, <a> elements
have their underline restored to make their link-ness more obvious.

aoeixsz4 commented on scoreboard/migrations/0001_initial.py:
```diff
@@ -1,5 +1,6 @@
-# Generated by Django 3.2.7 on 2021-10-08 14:55
```
Should we remove this and stop tracking anything in migrations or, if we're deliberately keeping this what was the reason?

copperwater replied:
I am fine with gitignoring migrations

aoeixsz4 commented on tnnt/views.py:
```diff
@@ -1,4 +1,18 @@
+from . import hardfought_utils # find_player
+
+def impossible(*args):
+    # Yes, I made impossible(). I'm too used to having it around.
```
I like having this impossible() but would suggest maybe having this additionally call die/exit or whatever, so that it behaves like a panic.
Also, there in the comment you say it's useful for seeing where something “really is known to be impossible” which I would reword as something like “catching things that *should* be impossible but aren't.”

copperwater replied:
I don't think we should crash the server and bring it down for *everyone* when someone manages to trigger one of these impossible cases (some of which can intentionally be done with a malicious POST request).

aoeixsz4 replied:
I assume it uses some async tasks and multithreading under the hood and only will crash the page, but I have yet to test this for sure.
In fact, at least in development-mode errors in code get logged to the terminal output and a detailed error page is displayed in the browser. When running in production however, the user won't be served this detailed stuff. Server doesn't totally crash tho. Then again deliberately calling die() or exit() maybe WOULD kill the server. Instead, I suggest we throw an exception.

copperwater committed:
Updates to clan management: clan size cap, fix a couple links

aoeixsz4 commented on tnnt/templates/clanmgmt.html:
```diff
@@ -0,0 +1,125 @@
+    <h3>{{clan.name}}</h3>
+    <form method="post" class="inline-form">
+      {% csrf_token %}
+        <button name="leave" type="submit" {% if members.count == 1 %}disabled{% endif %}>
```
Wouldn't we want to remove the clan if the last player left?

copperwater replied:
Submitting a "leave" POST request when the player is the sole clan member will indeed disband the clan (it just calls the disband_clan function), but in the UI there's a big "Disband Clan" button anyway, so there's no reason to confusingly have "Leave" up there as well.

aoeixsz4 replied:
Ahhhh okay, I gotchu.

aoeixsz4 commented on tnnt/views.py:
```diff
@@ -11,3 +26,310 @@ class AboutView(TemplateView):
+            invitee = hardfought_utils.find_player(request.POST['invitee'])
+            invitee.invites.add(player.clan)
+        except Player.DoesNotExist:
+            ctx['errmsg'] = 'No such player "' + request.POST['invitee'] + '" exists.'
```
Not sure if they can do any actual damage with this, but I don't like unvalidated input being concatenated into the error message like this.

aoeixsz4 commented on tnnt/views.py:
```diff
@@ -11,3 +26,310 @@ class AboutView(TemplateView):
+    # Helper function triggered when a create_clan POST request comes in
+    def create_clan(self, request, player, ctx):
+        create_clan_form = CreateClanForm(request.POST)
+        new_clan_name = request.POST['clan_name']
```
This will error if a user POSTs with 'create_clan' but no 'clan_name' in the request.

deepy replied:
shouldn't this use `create_clan_form.cleaned_data[]` after checking that the form `.is_valid()`?

aoeixsz4 replied:
not sure lol

copperwater replied:
done. perhaps @deepy can recommend how to make all the other forms on the page proper ones in forms.py, and make them validate by default as well, as that's something I just didn't have time for.

aoeixsz4 commented on tnnt/views.py:
```diff
@@ -11,3 +26,310 @@ class AboutView(TemplateView):
+    def rescind_invite(self, request, player, ctx):
+        # TODO? all of these helpers should test that the key is in POST, and if
+        # not, log an error and return
+        rescindee_id = request.POST['rescind_id']
```
Will error on invalid POST data from funny hacker user.

aoeixsz4 commented on tnnt/views.py:
```diff
@@ -11,3 +26,310 @@ class AboutView(TemplateView):
+
+    # Helper function triggered when "Make Admin" is clicked on a clan member
+    def make_admin(self, request, player, ctx):
+        new_admin_id = request.POST['kick_or_admin_id']
```
Will error on bad POST data.

aoeixsz4 commented on tnnt/views.py:
```diff
@@ -11,3 +26,310 @@ class AboutView(TemplateView):
+
+    # Helper function triggered when "Kick" is clicked on a clan member
+    def kick_member(self, request, player, ctx):
+        kickee_id = request.POST['kick_or_admin_id']
```
Will error on bad POST data.

aoeixsz4 commented on tnnt/views.py:
```diff
@@ -11,3 +26,310 @@ class AboutView(TemplateView):
+            return
+
+        # other player must be in this clan
+        if player.clan != kickee.clan:
```
I would use identity rather than equality here.

copperwater replied:
Using `if player.clan is not kickee.clan:` results in the admin not being able to kick anyone.

aoeixsz4 replied:
oh weird, nvm then

copperwater committed:
Enshrine some of aoei's feedback in FUTURE TODO comments

Things that could probably be improved, but are not a priority for the
2021 tournament.

copperwater committed:
Ignore migration files and remove existing one from version control

TNNT doesn't have long-term persistence; it doesn't need to run all
year and may well have different database models next year. Thus we're
not particularly interested in tracking database migrations or
constraining models.py changes to be backwards compatible with previous
database migrations.

copperwater committed:
Implement clan freeze

I didn't realize at the time that this is handled on the backend, but of
course it is. This is a configurable time (CLAN_FREEZE_TIME) after which
joining, inviting, and creating clans cannot be done. By default and
tradition, it's Nov 10 at midnight.

copperwater committed:
Fix a few clan management bugs

1. leave_clan didn't actually check that the player was in a clan
2. making admin and kicking were still referencing an "otherplayer"
   variable that didn't exist in those functions and would have crashed
3. aoei suggested not passing raw unvalidated POST data through to
   errmsg
4. some error messages still ended with punctuation; de facto standard
   for these is that they shouldn't

deepy commented on tnnt/views.py:
```diff
@@ -11,3 +25,305 @@ class AboutView(TemplateView):
+        # clan id actually needs to exist
+        # this could theoretically happen normally, where one player loads the
+        # clan management page and has an invite for another clan, which has
+        # disbanded by the time they submit the join request.
```
shouldn't the delete for the clan cascade and delete the invites as well?

copperwater replied:
Should happen automatically by django's managing of ManyToManyField relationships. (Confirmed by querying the database directly what happens if I create a clan, invite a bunch of people, and then disband it -- the scoreboard_player_invites table is left the same as it was before I made the clan.)

deepy replied:
You'd have to specify `on_delete` but having deletes cascade is generally the preferred solution

copperwater committed:
Use form.cleaned_data[] instead of request.POST for the two forms

copperwater committed:
Sort player lists on clan management page

Clan members by admin status first, then alphabetized. Invites just
alphabetized.

copperwater committed:
Add proper logging

Rather than a homegrown impossible() function that logged only to the
console, or print statements, this commit logs all clan management
events properly. I was noticing that I was trying to classify different
prints as different things - logging's multiple levels made sense for
this.

Most "impossible" cases - the majority of them actually possible if
clicking links in an outdated page or sending POST requests manually -
generate warnings. Normal clan events going through generate info
messages.

copperwater committed:
remove stray print statement in hdf auth backend

aoeixsz4 commented on tnnt/hardfought_utils.py:
```diff
@@ -0,0 +1,79 @@
+            print('ERROR connecting to or executing SQL on sqlite database')
+            print('  db path:', DGL_DATABASE_PATH)
+            print('  username:', username)
+        if uname is None:
```
This behaviour will mean I need a dummy dgl database for testing that has one each of everyone's nick...
I suppose I can generate such a thing in a similar way to your alice, bob, carol, dan one?

aoeixsz4 reviewed (approved):
ship it
